### PR TITLE
Bring XGBoost back as an optional dependency

### PR DESCRIPTION
**Pull request by KhaoticMind**
## What does this PR do?
The objective of this PR is to put XGBoost back in TPOT as an Optional dependency, as discussed on issue #154 

## Where should the reviewer start?
I've brought the documentation back from prior commits and added:
- checks to see if xgboost is present
- tests to new xgboost parameters
- ability to export xgboost operator that are part of a pipeline

## How should this PR be tested?
All nose tests (including new ones) are ok on my machine.
Tried running a toy test with just _xgb_boosting as a primitive in TPOT._pset and all was fine.

```
from tpot import TPOT
from sklearn.datasets import load_digits
from sklearn.cross_validation import train_test_split

digits = load_digits()
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.75, test_size=0.25)

tpot = TPOT(generations=3, verbosity=2, population_size=5)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
tpot.export('tpot_mnist_pipeline.py')
```

## Any background context you want to provide?

Not really...

## What are the relevant issues?

Issue in #154

### CODE REVIEW

**Comment by teaearlgraycold** [comment]
I'd suggest doing this instead:

And then in `__init__()`

```Python
try:
    from xgboost import XGBClassifier
    self._pset.addPrimitive(self._xg_boosting, [pd.DataFrame, float, float, int, int], pd.DataFrame)
except ImportError:
    self._pset.addPrimitive(self._gradient_boosting, [pd.DataFrame, float, float, float], pd.DataFrame)
```

Parameters for `_xg_boosting` might need tweaking, too.

**Comment by KhaoticMind** [comment]
Hi @teaearlgraycold. 
1) I kept _gradient_boosting and _xg_boosting because I was under the impression that this was the intent. But this is easy to overcome. Just say the word.
2) I prefer to keep the import at the top of the file, and use a flag do detect error, to respect pep8
3) I'll update the setup.py to add an extra_requires flag, so to make xgboost explicitly optional.
4) I'll update the travis scripts to depend on XGBoost

Please comment on 1 and 2.
I'll push new commits for 3 and 4 soon.

File: setup.py [hunk 1]
```diff
@@ -45,4 +45,5 @@ def calculate_version():
         'Topic :: Scientific/Engineering :: Artificial Intelligence'
     ],
     keywords=['pipeline optimization', 'hyperparameter optimization', 'data science', 'machine learning', 'genetic programming', 'evolutionary computation'],
+    extra_require = {'xgboost': ['xgboost']},
```
Comment (rhiever) [inline setup.py]: I believe it's `extras_require`: http://peak.telecommunity.com/DevCenter/setuptools#declaring-extras-optional-features-with-their-own-dependencies

File: tpot/export_utils.py [hunk 2]
```diff
@@ -393,6 +394,24 @@ def replace_function_calls(pipeline_list):
 
 {OUTPUT_DF}['gbc{OPERATOR_NUM}-classification'] = gbc{OPERATOR_NUM}.predict({OUTPUT_DF}.drop('class', axis=1).values)
 """.format(OUTPUT_DF=result_name, OPERATOR_NUM=operator_num, LEARNING_RATE=learning_rate, MAX_FEATURES=max_features, MIN_WEIGHT=min_weight)
+        
+        elif operator_name == '_xg_boosting':
+            
+            learning_rate = min(1., max(float(operator[3]), 0.0001))
+            subsample = min(1., max(0.1, float(operator[4])))            
+            min_child_weight = min(100, max(0, int(operator[5])))
+            max_depth = min(100, max(1, int(operator[6])))
```
Comment (rhiever) [inline tpot/export_utils.py]: How did you decide these parameters? Do you know these parameters have a significant effect, and are capped in the right range? (Just curious, as I'm not terribly familiar with XGBoost in practice.)

File: tpot/export_utils.py [hunk 3]
```diff
@@ -393,6 +394,24 @@ def replace_function_calls(pipeline_list):
...
+            min_child_weight = min(100, max(0, int(operator[5])))
+            max_depth = min(100, max(1, int(operator[6])))            
+
+            if result_name != operator[2]:
+                operator_text += "\n{OUTPUT_DF} = {INPUT_DF}.copy()".format(OUTPUT_DF=result_name, INPUT_DF=operator[2])
+
+            operator_text += """
+# Perform classification with a extreme gradient boosting classifier
+xgbc{OPERATOR_NUM} = XGBClassifier(learning_rate={LEARNING_RATE}, subsample={SUBSAMPLE}, min_child_weight={MIN_WEIGHT}, max_depth={MAX_DEPTH}, n_estimators=500)
+xgbc{OPERATOR_NUM}.fit({OUTPUT_DF}.loc[training_indices].drop('class', axis=1).values, {OUTPUT_DF}.loc[training_indices, 'class'].values)
```
Comment (rhiever) [inline tpot/export_utils.py]: Is there a tradeoff between `n_estimators` and e.g. `learning_rate` in a XGBClassifer? Or is it a good assumption that `n_estimators=500` is a good setting?

Comment (KhaoticMind) [inline tpot/export_utils.py]: They are closely related. A bigger learning rate usually requires a lower n_estimators to prevent overfitting.
For the xgboost Python interface the default values are learning_rate=0.1 and n_estimators=100 (http://xgboost.readthedocs.io/en/latest/python/python_api.html#module-xgboost.sklearn).

I limited n_estimators here just as it has been done with GrandientBoostingClassifier (where learning_rate / n_estimators play a similar (equal?) role).

It might be a good idea to let DEAP search for n_estimators as well as learning_rate. If this is the case, it should also be done for the GrandientBoostingClassifier.

File: tpot/export_utils.py (hunk 2 above)
Comment (KhaoticMind) [inline tpot/export_utils.py]: I've used learning_rate, subsample and min_child_weight as these are the counterparts of     learning_rate, max_features and min_weight present in _gradient_boosting. Also, from my experience, max_depth plays a very important role to prevent overfitting (https://github.com/dmlc/xgboost/blob/master/doc/how_to/param_tuning.md).

I got the value caps from https://github.com/dmlc/xgboost/blob/master/doc/parameter.md (learning_rate (eta) = [0,1], subsample = (0, 1], min_child_weight = [0, inifity], max_depth = [1, inifity]).
Base on these I used some common sense for:
-  subsample: 0 samples per tree makes no sense for me
- max_depth : a depth of 0 means no tree (for what I understand), and a tree that is too deep will surely overfit

File: setup.py (hunk 1 above)
Comment (KhaoticMind) [inline setup.py]: You are right. Fixed it on 2910acd.

File: tpot/export_utils.py (hunk 3 above)
Comment (rhiever) [inline tpot/export_utils.py]: IIRC GradientBoostingClassifier wasn't tremendously affected by `n_estimators` in our sklearn benchmarks. Maybe we need to look at that again.

File: tpot/export_utils.py (hunk 2 above)
Comment (rhiever) [inline tpot/export_utils.py]: Makes sense. If we have `min_child_weight`, then there is no need for `max_depth`. They both do the same thing (limit tree depth), albeit through different mechanisms.

In cases where a number is specified (e.g. `min_child_weight` seems to be a count?), is it possible to pass a float indicating a fraction instead?

Comment (KhaoticMind) [inline tpot/export_utils.py]: Yeah. Haven't thought of that. We can keep just `min_child_weight` and make `max_depth` a really large  value (or `None` if this is accept by `xgboost`) to see if this works. I'll work on it.

For `min_child_weight`, (for what I know) it can't be a fraction meaning a fraction of the data. On `xgboost` the classes can have weights (created on the `DMatrix` datatype used internally by it), so it accepts fractions here just to reflect that some classes might have fractional weight.

In the case of the `sklearn` interface (`XGBClassifier`) all classes should have weight 1. So the min value here should be 1. I'll also fix this.

File: tpot/export_utils.py (hunk 3 above)
Comment (KhaoticMind) [inline tpot/export_utils.py]: Take a look here: http://pt.slideshare.net/DataRobot/gradient-boosted-regression-trees-in-scikitlearn
The parameter tuning section starts on slide 22. The effect of `learning_rate` and `n_estimators` can be seen on slide 25.

**Comment by KhaoticMind** [comment]
Some more commits to fix some pending issues discussed above.
Please feel free to comment on this new version.

**Comment by rhiever** [comment]
Hi @KhaoticMind, thank you so much for working XGBoost back into TPOT as an optional dependency! As you may have seen, we've been refactoring TPOT so it's easier to maintain. You can find the latest version of the refactor in the [development branch](https://github.com/rhiever/tpot/tree/development).

This change will require some rework of your PR. Would you prefer to rework the PR into the new framework, or would you like @teaearlgraycold and I to take your code and rework it?

[Also note: all PRs must be sent to the `development` branch now, as the `master` branch is intended for the latest stable version of TPOT.]

**Comment by KhaoticMind** [comment]
Hello @rhiever! I think I'll prefer to leave it to @teaearlgraycold , as I'm currently short on time :(
I'll keep an eye on the project, and as soon as I have more spare time I'll head back and try to make some more contributions.

**Comment by rhiever** [comment]
No problem! Thank you for your contributions. :-)
